vendors/swedbank-pay-core-library/src/SwedbankPay/Core/Library/OrderAction.php

```diff
@@ -0,0 +1,573 @@
+            ]
+        ];
+        $result = $this->request('POST', $href, $params);
+
+        // Save transaction
+        $transaction = $result['reversal']['transaction'];
+        $this->saveTransaction($orderId, $transaction);
+
+        switch ($transaction['state']) {
+            case 'Completed':
+                $this->updateOrderStatus(
+                    OrderInterface::STATUS_REFUNDED,
+                    sprintf('Refunded: %s.', $amount)
+                );
+                break;
+            case 'Initialized':
+            case 'AwaitingActivity':
+                $this->updateOrderStatus(OrderInterface::STATUS_CANCELLED,
+                    sprintf('Transaction reversal status: %s.', $transaction['state'])
+                );
+                break;
+            case 'Failed':
+                $message = isset($transaction['failedReason']) ? $transaction['failedReason'] : 'Refund is failed.';
+                throw new Exception($message);
+            default:
+                throw new Exception('Refund is failed.');
+        }
+
+        return $result;
+    }
+
+    /**
+     * Abort Payment.
+     *
+     * @param mixed $orderId
+     *
+     * @return Response
+     * @throws Exception
+     */
+    public function abort($orderId)
+    {
+        /** @var Order $order */
+        $order = $this->getOrder($orderId);
+
+        $paymentId = $order->getPaymentId();
+        if (empty($paymentId)) {
+            throw new Exception('Unable to get payment ID');
+        }
+
+        // @todo Check if order has been paid
+        $href = $this->fetchPaymentInfo($paymentId)->getOperationByRel('update-payment-abort');
+        if (empty($href)) {
+            throw new Exception('Abort is unavailable');
+        }
+
+        $params = [
+            'payment' => [
+                'operation' => 'Abort',
+                'abortReason' => 'CancelledByConsumer'
+            ]
+        ];
+        $result = $this->request('PATCH', $href, $params);
+
+        if ($result['payment']['state'] === 'Aborted') {
+            $this->updateOrderStatus(OrderInterface::STATUS_CANCELLED, 'Payment aborted');
+        } else {
+            throw new Exception('Aborting is failed.');
+        }
+
+        return $result;
+    }
+
+    /**
+     * Update Order Status.
+     *
+     * @param mixed $orderId
+     * @param string $status
+     * @param string|null $message
+     * @param string|null $transactionId
+     */
+    public function updateOrderStatus($orderId, $status, $message = null, $transactionId = null)
+    {
+        $this->adapter->updateOrderStatus($orderId, $status, $message, $transactionId);
+    }
+
+    /**
+     * Analyze the transaction and update the related order.
+     *
+     * @param $orderId
+     * @param Transaction|array $transaction
+     *
+     * @throws Exception
+     */
+    public function processTransaction($orderId, $transaction)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method processTransaction() has 162 lines of code. Current threshold is set to 100. Avoid really long methods.](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-woocommerce-payments/pullRequest?prid=5340032)